File: src/lib/androidDownload.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { isAndroidClient } from "./platform";
import { syncHistoryFromQueueItems } from "./historySync";
import { useQueueStore } from "../store/queue";
import { useSettingsStore } from "../store/settings";
import type { AppSettings, QueueItem, QueueStatus } from "../types";
import { normalizeQueueError } from "./errors";

export interface OrchestratorSyncPayload {
  items: QueueItem[];
  settings: Partial<AppSettings>;
  activeId: string | null;
  globalBotStreak?: number;
  successCount?: number;
  pausedUntil?: number;
}

export interface NotificationPermissionResult {
  granted: boolean;
  requested?: boolean;
}

export interface PendingNavigation {
  sharedUrl?: string | null;
  openPage?: string | null;
}

export interface YtdlpUpdateInfo {
  currentVersion: string;
  latestVersion: string;
  updateAvailable: boolean;
  downloadUrl?: string | null;
  sha256?: string | null;
}

export interface ApplyYtdlpResult {
  success: boolean;
  version?: string;
  error?: string | null;
}

/** Fired when orchestrator merge changes item status (for toast notifications). */
export type MergeTransition = {
  id: string;
  from: QueueStatus;
  to: QueueStatus;
  item: QueueItem;
};

let onMergeTransitions: ((transitions: MergeTransition[]) => void) | null =
  null;

/** Register handler for status transitions after native queue merge. */
export function setOrchestratorMergeNotifier(
  fn: ((transitions: MergeTransition[]) => void) | null,
): void {
  onMergeTransitions = fn;
}

/** Native orchestrator user-pause flag (synced from queue-state JSON). */
let orchestratorPaused = false;
let orchestratorPausedUntil = 0;

export function isOrchestratorPaused(): boolean {
  return orchestratorPaused;
}

/** Global orchestrator cooldown (guest backoff / safe bulk). */
export function getOrchestratorPausedUntil(): number {
  return orchestratorPausedUntil;
}

/** Start foreground download service and restore orchestrator. */
export async function ensureDownloadService(): Promise<void> {
  return invoke("ensure_download_service");
}

/** Sync React queue to Kotlin DownloadOrchestrator. */
export async function syncOrchestratorQueue(
  payload: OrchestratorSyncPayload,
): Promise<void> {
  return invoke("sync_orchestrator_queue", { payload });
}

/** Read native orchestrator state JSON. */
export async function getOrchestratorState(): Promise<string> {
  return invoke<string>("get_orchestrator_state");
}

export async function pauseActiveJob(): Promise<void> {
  return invoke("pause_active_job");
}

export async function resumeOrchestratorJob(id: string): Promise<void> {
  return invoke("resume_orchestrator_job", { id });
}

export async function retryOrchestratorJob(id: string): Promise<void> {
  return invoke("retry_orchestrator_job", { id });
}

export async function pauseOrchestrator(): Promise<void> {
  return invoke("pause_orchestrator");
}

export async function resumeOrchestrator(): Promise<void> {
  return invoke("resume_orchestrator");
}

export async function cancelOrchestratorJob(id: string): Promise<void> {
  return invoke("cancel_orchestrator_job", { id });
}

export async function requeueFailedBotItems(): Promise<void> {
  return invoke("requeue_failed_bot_items");
}

export async function requestNotificationPermission(): Promise<NotificationPermissionResult> {
  return invoke<NotificationPermissionResult>("request_notification_permission");
}

export async function getPendingNavigation(): Promise<PendingNavigation> {
  return invoke<PendingNavigation>("get_pending_navigation");
}

export async function openYoutubeLogin(): Promise<void> {
  return invoke("open_youtube_login");
}

/** Write a debug line to Android logcat (no-op on desktop). */
export async function logAndroid(tag: string, message: string): Promise<void> {
  if (!isAndroidClient()) return;
  try {
    await invoke("android_log", { tag, message });
  } catch {
    console.log(`[Acorn/${tag}] ${message}`);
  }
}

export async function checkYtdlpUpdate(force = false): Promise<YtdlpUpdateInfo> {
  await logAndroid("YtdlpBridge", `checkYtdlpUpdate invoke force=${force}`);
  const info = await invoke<YtdlpUpdateInfo>("check_ytdlp_update", { force });
  await logAndroid(
    "YtdlpBridge",
    `checkYtdlpUpdate result current=${info.currentVersion} latest=${info.latestVersion} available=${info.updateAvailable}`,
  );
  return info;
}

export async function applyYtdlpUpdate(): Promise<ApplyYtdlpResult> {
  await logAndroid("YtdlpBridge", "applyYtdlpUpdate invoke");
  const result = await invoke<ApplyYtdlpResult>("apply_ytdlp_update");
  await logAndroid(
    "YtdlpBridge",
    `applyYtdlpUpdate result success=${result.success} version=${result.version ?? ""} error=${result.error ?? ""}`,
  );
  return result;
}

export async function openBatteryOptimizationSettings(): Promise<void> {
  return invoke("open_battery_optimization_settings");
}

/** Fingerprint queue composition — excludes status so native progress is not overwritten. */
export function buildOrchestratorSyncKey(
  items: QueueItem[],
  settings: Partial<AppSettings>,
): string {
  const itemKey = items
    .map(
      (i) =>
        `${i.id}|${i.url}|${i.format}|${i.audioOnly}|${i.container}|${i.audioFormat}|${i.writeSubs}|${i.writeThumbnail}|${i.subLangs}`,
    )
    .join("\n");
  const settingsKey = [
    settings.cookiesFile ?? "",
    settings.androidDownloadJitter,
    settings.androidSafeBulkMode,
    settings.androidSafeBulkInterval,
    settings.androidCookieRotateInterval,
    settings.androidGuestPaceMode,
    settings.androidYtdlpAutoUpdate,
  ].join("|");
  return `${itemKey}::${settingsKey}`;
}

/** Native orchestrator item snapshot from Kotlin JSON. */
interface OrchestratorSnapshotItem {
  id: string;
  url: string;
  title?: string;
  status: QueueStatus;
  percent?: number;
  filename?: string | null;
  error?: string | null;
  retryCount?: number;
  lastBotErrorAt?: number;
  cooldownUntil?: number;
}

const STATUS_RANK: Record<QueueStatus, number> = {
  cancelled: 1,
  queued: 2,
  paused: 2,
  error: 3,
  downloading: 4,
  completed: 5,
};

const TERMINAL_STATUSES = new Set<QueueStatus>([
  "error",
  "completed",
  "cancelled",
]);

/** Pick the authoritative status when merging native snapshot with React store. */
function mergeQueueStatus(
  nativeStatus: QueueStatus,
  existingStatus: QueueStatus,
  nativeCooldown: number | undefined,
  now: number,
): QueueStatus {
  if (existingStatus === "cancelled") return "cancelled";
  if (nativeStatus === "paused") return "paused";
  if (existingStatus === "paused" && nativeStatus === "downloading") {
    return "paused";
  }
  if (
    nativeStatus === "queued" &&
    nativeCooldown != null &&
    nativeCooldown > now &&
    existingStatus === "downloading"
  ) {
    return "queued";
  }
  if (TERMINAL_STATUSES.has(nativeStatus)) return nativeStatus;
  if (
    TERMINAL_STATUSES.has(existingStatus) &&
    nativeStatus !== "downloading" &&
    nativeStatus !== "queued"
  ) {
    return existingStatus;
  }
  const nativeRank = STATUS_RANK[nativeStatus] ?? 0;
  const existingRank = STATUS_RANK[existingStatus] ?? 0;
  if (nativeRank > existingRank) return nativeStatus;
  if (existingRank > nativeRank) return existingStatus;
  return nativeStatus;
}

/** Build a minimal queue row when native has an item React never synced. */
function minimalQueueItemFromNative(
  oItem: OrchestratorSnapshotItem,
): QueueItem {
  const settings = useSettingsStore.getState();
  return {
    id: oItem.id,
    url: oItem.url,
    title: oItem.title ?? oItem.url,
    format: settings.defaultQuality,
    audioOnly: false,
    audioFormat: "mp3",
    container: settings.defaultContainer,
    writeSubs: false,
    subLangs: "tr",
    writeThumbnail: settings.writeThumbnailDefault ?? false,
    status: oItem.status,
    percent: oItem.percent ?? 0,
    filename: oItem.filename ?? undefined,
    error: normalizeQueueError(oItem.error),
    createdAt: Date.now(),
    retryCount: oItem.retryCount,
    lastBotErrorAt: oItem.lastBotErrorAt,
    cooldownUntil: oItem.cooldownUntil,
  };
}

/** Merge Kotlin orchestrator snapshot into the React queue store (upsert by id). */
export function mergeOrchestratorState(stateJson: string): void {
  try {
    const state = JSON.parse(stateJson) as {
      items?: OrchestratorSnapshotItem[];
      activeId?: string | null;
      paused?: boolean;
      pausedUntil?: number;
    };
    orchestratorPaused = state.paused === true;
    orchestratorPausedUntil = state.pausedUntil ?? 0;
    useQueueStore.getState().setOrchestratorPaused(orchestratorPaused);
    const store = useQueueStore.getState();
    const nativeItems = state.items ?? [];

    if (nativeItems.length === 0) {
      if (state.activeId !== undefined) {
        store.setActiveId(state.activeId);
      }
      void logAndroid("QueueMerge", "empty native snapshot");
      return;
    }

    const existingById = new Map(store.items.map((item) => [item.id, item]));
    const nativeIds = new Set(nativeItems.map((item) => item.id));
    const now = Date.now();
    let updated = 0;
    let inserted = 0;

    const mergedNative = nativeItems.map((oItem) => {
      const existing = existingById.get(oItem.id);
      if (existing) {
        updated += 1;
        const nativePercent = oItem.percent ?? 0;
        const existingPercent = existing.percent ?? 0;
        return {
          ...existing,
          title: oItem.title ?? existing.title,
          url: oItem.url || existing.url,
          status: mergeQueueStatus(
            oItem.status,
            existing.status,
            oItem.cooldownUntil,
            now,
          ),
          percent: Math.max(nativePercent, existingPercent),
          speed: existing.speed,
          eta: existing.eta,
          phase:
            oItem.status === "completed" ||
            oItem.status === "error" ||
            oItem.status === "cancelled"
              ? null
              : existing.phase,
          filename: oItem.filename ?? existing.filename,
          error:
            oItem.status === "downloading" || oItem.status === "queued"
              ? normalizeQueueError(oItem.error)
              : normalizeQueueError(oItem.error) ?? normalizeQueueError(existing.error),
          retryCount: oItem.retryCount ?? existing.retryCount,
          lastBotErrorAt: oItem.lastBotErrorAt ?? existing.lastBotErrorAt,
          cooldownUntil: oItem.cooldownUntil ?? existing.cooldownUntil,
        };
      }
      inserted += 1;
      return minimalQueueItemFromNative(oItem);
    });

    const reactOnlyQueued = store.items.filter(
      (item) => !nativeIds.has(item.id) && item.status === "queued",
    );
    const merged = [...mergedNative, ...reactOnlyQueued];

    const prevById = new Map(store.items.map((item) => [item.id, item]));
    const transitions: MergeTransition[] = [];
    for (const item of merged) {
      const prev = prevById.get(item.id);
      if (prev && prev.status !== item.status) {
        transitions.push({
          id: item.id,
          from: prev.status,
          to: item.status,
          item,
        });
      }
    }

    useQueueStore.setState({
      items: merged,
      activeId: state.activeId ?? store.activeId,
    });

    if (transitions.length > 0) {
      onMergeTransitions?.(transitions);
    }

    syncHistoryFromQueueItems(merged);

    void logAndroid(
      "QueueMerge",
      `upsert total=${nativeItems.length} updated=${updated} inserted=${inserted} active=${state.activeId ?? "none"}`,
    );
  } catch (err) {
    void logAndroid(
      "QueueMerge",
      `parse failed: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
}

/** Pull native orchestrator state and merge into React store. */
export async function pullOrchestratorState(): Promise<void> {
  if (!isAndroidClient()) return;
  try {
    const json = await getOrchestratorState();
    mergeOrchestratorState(json);
  } catch (err) {
    void logAndroid(
      "QueueMerge",
      `pull failed: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
}

/** Subscribe to native orchestrator queue-state events. */
export async function onQueueState(
  handler: (stateJson: string) => void,
): Promise<UnlistenFn> {
  return listen<{ state: string }>("queue-state", (event) => {
    void logAndroid("QueueMerge", "queue-state event received");
    handler(event.payload.state);
  });
}

File: src-tauri/src/android_download.rs
//! Android download service + orchestrator bridge (DownloadPlugin.kt).

#[cfg(target_os = "android")]
use crate::download::adapter::build_android_base_download_args;
use crate::download::ytdlp::DownloadOptions;
use serde::{Deserialize, Serialize};
use tauri::{
    plugin::{Builder, TauriPlugin},
    AppHandle, Runtime,
};

#[cfg(target_os = "android")]
use tauri::Manager;

#[cfg(target_os = "android")]
struct DownloadPluginHandle<R: Runtime>(tauri::plugin::PluginHandle<R>);

/// Queue item payload from React for orchestrator sync.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OrchestratorSyncItem {
    pub id: String,
    pub url: String,
    pub title: String,
    pub status: String,
    #[serde(default)]
    pub percent: f64,
    pub format: String,
    pub audio_only: bool,
    pub audio_format: String,
    pub container: String,
    pub write_subs: bool,
    pub sub_langs: String,
    pub write_thumbnail: bool,
    #[serde(default)]
    pub retry_count: i32,
    #[serde(default)]
    pub last_bot_error_at: i64,
    #[serde(default)]
    pub cooldown_until: i64,
    #[serde(default)]
    pub filename: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
}

/// Full orchestrator sync payload from React store.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OrchestratorSyncPayload {
    pub items: Vec<OrchestratorSyncItem>,
    pub settings: serde_json::Value,
    #[serde(default)]
    pub active_id: Option<String>,
    #[serde(default)]
    pub global_bot_streak: i32,
    #[serde(default)]
    pub success_count: i32,
    #[serde(default)]
    pub paused_until: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct NotificationPermissionResult {
    pub granted: bool,
    #[serde(default)]
    pub requested: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct PendingNavigation {
    #[serde(default)]
    pub shared_url: Option<String>,
    #[serde(default)]
    pub open_page: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct YtdlpUpdateInfo {
    pub current_version: String,
    pub latest_version: String,
    pub update_available: bool,
    #[serde(default)]
    pub download_url: Option<String>,
    #[serde(default)]
    pub sha256: Option<String>,
}

/// Register Kotlin DownloadPlugin.
pub fn init<R: Runtime>() -> TauriPlugin<R> {
    Builder::new("android-download")
        .setup(|app, api| {
            #[cfg(target_os = "android")]
            {
                let handle =
                    api.register_android_plugin("com.acorn.videodownloader", "DownloadPlugin")?;
                app.manage(DownloadPluginHandle(handle));
            }
            #[cfg(not(target_os = "android"))]
            {
                let _ = (app, api);
            }
            Ok(())
        })
        .build()
}

#[cfg(target_os = "android")]
async fn resolve_android_cookies<R: Runtime>(app: &AppHandle<R>, cookies_file: Option<&str>) -> Option<String> {
    if let Some(path) = cookies_file.filter(|p| !p.trim().is_empty()) {
        return Some(path.to_string());
    }
    crate::android_ytdlp::get_cookies_path(app).await.ok()
}

#[cfg(target_os = "android")]
fn sync_item_to_download_options(item: &OrchestratorSyncItem, settings: &serde_json::Value) -> DownloadOptions {
    DownloadOptions {
        id: item.id.clone(),
        url: item.url.clone(),
        format: if item.audio_only {
            "bestaudio/best".to_string()
        } else {
            item.format.clone()
        },
        output_dir: settings
            .get("outputDir")
            .and_then(|v| v.as_str())
            .unwrap_or("downloads")
            .to_string(),
        audio_only: item.audio_only,
        audio_format: Some(item.audio_format.clone()),
        container: Some(item.container.clone()),
        write_subs: item.write_subs,
        sub_langs: Some(item.sub_langs.clone()).filter(|s| !s.is_empty()),
        write_thumbnail: item.write_thumbnail,
        ytdlp_path: None,
        ffmpeg_path: None,
        cookies_from_browser: None,
        cookies_file: settings
            .get("cookiesFile")
            .and_then(|v| v.as_str())
            .map(str::to_string),
    }
}

#[cfg(target_os = "android")]
fn build_orchestrator_json(
    payload: &OrchestratorSyncPayload,
    options_by_id: &[(String, Vec<Vec<String>>)],
) -> String {
    use serde_json::json;

    let options_map: std::collections::HashMap<_, _> = options_by_id.iter().cloned().collect();
    let items: Vec<_> = payload
        .items
        .iter()
        .map(|item| {
            let options = options_map.get(&item.id).cloned().unwrap_or_default();
            json!({
                "id": item.id,
                "url": item.url,
                "title": item.title,
                "status": item.status,
                "percent": item.percent,
                "retryCount": item.retry_count,
                "lastBotErrorAt": item.last_bot_error_at,
                "cooldownUntil": item.cooldown_until,
                "filename": item.filename,
                "error": item.error,
                "options": options,
            })
        })
        .collect();

    serde_json::json!({
        "items": items,
        "settings": payload.settings,
        "activeId": payload.active_id,
        "globalBotStreak": payload.global_bot_streak,
        "successCount": payload.success_count,
        "pausedUntil": payload.paused_until,
    })
    .to_string()
}

/// Start foreground service and restore orchestrator state.
#[tauri::command]
pub async fn ensure_download_service<R: Runtime>(app: AppHandle<R>) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("ensureDownloadService", ())
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok(())
    }
}

/// Sync React queue snapshot to Kotlin DownloadOrchestrator.
#[tauri::command]
pub async fn sync_orchestrator_queue<R: Runtime>(
    app: AppHandle<R>,
    payload: OrchestratorSyncPayload,
) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let cookies = resolve_android_cookies(
            &app,
            payload
                .settings
                .get("cookiesFile")
                .and_then(|v| v.as_str()),
        )
        .await;

        let mut options_by_id = Vec::new();
        for item in &payload.items {
            let opts = sync_item_to_download_options(item, &payload.settings);
            let flat = build_android_base_download_args(&opts, cookies.as_deref());
            let options = crate::android_ytdlp::flat_args_to_options(&flat);
            options_by_id.push((item.id.clone(), options));
        }

        let json = build_orchestrator_json(&payload, &options_by_id);
        #[derive(Serialize)]
        struct SyncArgs {
            json: String,
        }
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("syncOrchestratorQueue", SyncArgs { json })
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = (app, payload);
        Ok(())
    }
}

/// Read orchestrator state JSON from Kotlin.
#[tauri::command]
pub async fn get_orchestrator_state<R: Runtime>(app: AppHandle<R>) -> Result<String, String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let result: serde_json::Value = state
            .0
            .run_mobile_plugin_async("getOrchestratorState", ())
            .await
            .map_err(|e| e.to_string())?;
        result
            .get("state")
            .and_then(|v| v.as_str())
            .map(str::to_string)
            .ok_or_else(|| "Orchestrator state missing".to_string())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok("{}".into())
    }
}

#[tauri::command]
pub async fn pause_orchestrator<R: Runtime>(app: AppHandle<R>) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("pauseOrchestrator", ())
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok(())
    }
}

#[tauri::command]
pub async fn resume_orchestrator<R: Runtime>(app: AppHandle<R>) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("resumeOrchestrator", ())
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok(())
    }
}

#[tauri::command]
pub async fn pause_active_job<R: Runtime>(app: AppHandle<R>) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("pauseActiveJob", ())
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok(())
    }
}

#[tauri::command]
pub async fn resume_orchestrator_job<R: Runtime>(
    app: AppHandle<R>,
    id: String,
) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        #[derive(Serialize)]
        struct IdArgs {
            id: String,
        }
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("resumeOrchestratorJob", IdArgs { id })
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = (app, id);
        Ok(())
    }
}

#[tauri::command]
pub async fn retry_orchestrator_job<R: Runtime>(
    app: AppHandle<R>,
    id: String,
) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        #[derive(Serialize)]
        struct IdArgs {
            id: String,
        }
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("retryOrchestratorJob", IdArgs { id })
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = (app, id);
        Ok(())
    }
}

#[tauri::command]
pub async fn cancel_orchestrator_job<R: Runtime>(app: AppHandle<R>, id: String) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        #[derive(Serialize)]
        struct IdArgs {
            id: String,
        }
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("cancelOrchestratorJob", IdArgs { id })
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = (app, id);
        Ok(())
    }
}

#[tauri::command]
pub async fn requeue_failed_bot_items<R: Runtime>(app: AppHandle<R>) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("requeueFailedBotItems", ())
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok(())
    }
}

#[tauri::command]
pub async fn request_notification_permission<R: Runtime>(
    app: AppHandle<R>,
) -> Result<NotificationPermissionResult, String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        state
            .0
            .run_mobile_plugin_async("requestNotificationPermission", ())
            .await
            .map_err(|e| e.to_string())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok(NotificationPermissionResult {
            granted: true,
            requested: false,
        })
    }
}

#[tauri::command]
pub async fn get_pending_navigation<R: Runtime>(
    app: AppHandle<R>,
) -> Result<PendingNavigation, String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        state
            .0
            .run_mobile_plugin_async("getPendingNavigation", ())
            .await
            .map_err(|e| e.to_string())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok(PendingNavigation::default())
    }
}

#[tauri::command]
pub async fn open_youtube_login<R: Runtime>(app: AppHandle<R>) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("openYoutubeLogin", ())
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Err("YouTube login is only available on Android.".into())
    }
}

#[tauri::command]
pub async fn check_ytdlp_update<R: Runtime>(
    app: AppHandle<R>,
    force: Option<bool>,
) -> Result<YtdlpUpdateInfo, String> {
    #[cfg(target_os = "android")]
    {
        #[derive(Serialize)]
        struct ForceArgs {
            force: bool,
        }
        let state = app.state::<DownloadPluginHandle<R>>();
        state
            .0
            .run_mobile_plugin_async(
                "checkYtdlpUpdate",
                ForceArgs {
                    force: force.unwrap_or(false),
                },
            )
            .await
            .map_err(|e| e.to_string())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Err("yt-dlp update is only available on Android.".into())
    }
}

#[tauri::command]
pub async fn apply_ytdlp_update<R: Runtime>(app: AppHandle<R>) -> Result<serde_json::Value, String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        state
            .0
            .run_mobile_plugin_async("applyYtdlpUpdate", ())
            .await
            .map_err(|e| e.to_string())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Err("yt-dlp update is only available on Android.".into())
    }
}

#[tauri::command]
pub async fn open_battery_optimization_settings<R: Runtime>(app: AppHandle<R>) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("openBatteryOptimizationSettings", ())
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = app;
        Ok(())
    }
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AndroidLogArgs {
    pub tag: String,
    pub message: String,
}

#[tauri::command]
pub async fn android_log<R: Runtime>(app: AppHandle<R>, tag: String, message: String) -> Result<(), String> {
    #[cfg(target_os = "android")]
    {
        let state = app.state::<DownloadPluginHandle<R>>();
        let _: serde_json::Value = state
            .0
            .run_mobile_plugin_async("androidLog", AndroidLogArgs { tag, message })
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = (app, tag, message);
        Ok(())
    }
}

/// Build yt-dlp option pairs for one Android download (used by orchestrator sync).
#[tauri::command]
pub async fn prepare_android_download_options<R: Runtime>(
    app: AppHandle<R>,
    opts: DownloadOptions,
) -> Result<Vec<Vec<String>>, String> {
    #[cfg(target_os = "android")]
    {
        let cookies = resolve_android_cookies(&app, opts.cookies_file.as_deref()).await;
        let flat = build_android_base_download_args(&opts, cookies.as_deref());
        Ok(crate::android_ytdlp::flat_args_to_options(&flat))
    }
    #[cfg(not(target_os = "android"))]
    {
        let _ = (app, opts);
        Ok(vec![])
    }
}
